warp10/src/main/java/io/warp10/script/functions/POLYFUNC.java

```diff
@@ -0,0 +1,171 @@
+    Object top = stack.pop();
+
+    if (top instanceof List) {
+      List<Object> list = new ArrayList<Object>(((List) top).size());
+      for (Object elt: (List) top) {
+        if (!(elt instanceof Number)) {
+          throw new WarpScriptException(getName() + " can only be applied to numerical values.");
+        }
+        list.add(this.func.value(((Number) elt).doubleValue()));
+      }
+      stack.push(list);
+    } else if (top instanceof GeoTimeSerie) {
+      GeoTimeSerie gts = (GeoTimeSerie) top;
+      int n = GTSHelper.nticks(gts);
+      GeoTimeSerie out = new GeoTimeSerie(n);
+      if (GTSHelper.isBucketized(gts)) {
+        long lastbucket = GTSHelper.getLastBucket(gts);
+        long bucketspan = GTSHelper.getBucketSpan(gts);
+        for (int i = 0; i < n; i++) {
+          long ts = lastbucket - i * bucketspan;
+          double value = this.func.value(ts);
+          GTSHelper.setValue(out, ts, GTSHelper.locationAtTick(gts, ts), GTSHelper.elevationAtTick(gts, ts), value, false);
+        }
+      } else {
+        for (int i = 0; i < n; i++) {
+          long ts = GTSHelper.tickAtIndex(gts, i);
+          double value = this.func.value(ts);
+          GTSHelper.setValue(out, ts, GTSHelper.locationAtIndex(gts, i), GTSHelper.elevationAtIndex(gts, i), value, false);
+        }
+      }
+      stack.push(out);
+    } else if (top instanceof GTSEncoder) {
+      GTSDecoder decoder = ((GTSEncoder) top).getDecoder();
+      GTSEncoder encoder = new GTSEncoder(decoder.getBaseTimestamp());
+      encoder.setMetadata(decoder.getMetadata());
+
+      while(decoder.next()) {
+        long ts = decoder.getTimestamp();
+        double value = this.func.value(ts);
+        encoder.addValue(ts, decoder.getLocation(), decoder.getElevation(), value);
+      }
+      stack.push(encoder);
+    } else if (top instanceof Number) {
+      double value = this.func.value(((Number) top).doubleValue());
+
+      stack.push(value);
+    } else if (!(top instanceof Number)) {
+      throw new WarpScriptException(getName() + " can only be applied to numerical values.");
+    }
+
+
+    return stack;
+  }
+
+  @Override
+  public Object apply(Object[] args) throws WarpScriptException {
+    long tick = (long) args[0];
+    long[] locations = (long[]) args[4];
+    long[] elevations = (long[]) args[5];
+    Object[] values = (Object[]) args[6];
+
+    if (0 == values.length) {
+      return new Object[] { tick, GeoTimeSerie.NO_LOCATION, GeoTimeSerie.NO_ELEVATION, this.func.value(tick) };
+    }
+
+    if (1 != values.length) {
+      throw new WarpScriptException(getName() + " can only be applied to a single value.");
+    }
+
+    return new Object[] { tick, locations[0], elevations[0], this.func.value(tick) };
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder sb = new StringBuilder();
+    try {
+      SNAPSHOT.addElement(sb, Arrays.asList(this.func.getCoefficients()));
```

Comment:
Same problem with Arrays.asList, prefer:

```
      double[] coeffs = this.func.getCoefficients();
      ArrayList<Double> coeffsList = new ArrayList<>(coeffs.length);
      for(double coeff: coeffs) {
        coeffsList.add(coeff);
      }

      SNAPSHOT.addElement(sb, coeffsList);
```

Reply:
What is the point of copying the list to create its SNAPSHOT?